pkg/proxy/usergating/controller.go

```diff
@@ -0,0 +1,195 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package usergating
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	kcpcache "github.com/kcp-dev/apimachinery/pkg/cache"
+	"github.com/kcp-dev/client-go/informers"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/proxy/usergating/ruleset"
+)
+
+const (
+	controllerName = "kcp-user-gating"
+	resyncPeriod   = 10 * time.Minute
+
+	UserGateNamespace = "kcp-system"
+	UserGateResource  = "usergate.kcp.dev"
+	RuleKey           = "ruleset.yaml"
+)
+
+type SecretGetter func(string) (*v1.Secret, error)
+
+// Controller maintains a set of rules configured from a secret in
+// root:kcp-system that determine which users can use the service.
+type Controller struct {
+	queue workqueue.RateLimitingInterface
+
+	GetSecret SecretGetter
+
+	lock  sync.RWMutex
+	Rules *ruleset.RuleSet
+
+	DefaultRules *ruleset.RuleSet
+}
+
+// NewController creates a controller that watches a secret named
+// usergate.kcp.dev in root:kcp-system. The secret contains yaml under the key
+// "usergate.yaml" that describes which users are allowed or denied access to
+// the system.
+func NewController(ctx context.Context, versionedInformers informers.SharedInformerFactory, defaultRules *ruleset.RuleSet) *Controller {
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName)
+	secretInformer := versionedInformers.Core().V1().Secrets().Cluster(tenancyv1alpha1.RootCluster)
+
+	c := &Controller{
+		queue:        queue,
+		GetSecret:    secretInformer.Lister().Secrets(UserGateNamespace).Get,
```

Comment:
a proxy allowed to read root secrets make me nervous.

Reply:
agreed and afais this is not necessary here.

Reply:
We could (should?) have some other workspace or at least namespace for it.